lib/client.js

```diff
@@ -124,10 +124,13 @@ function Client (broker, conn, req) {
         if (clientSub && (clientSub.qos || 0) < packet.qos) {
           packet.qos = clientSub.qos
         }
-        packet.writeCallback = cb
         if (that.clean || packet.retain) {
+          packet.writeCallback = cb
           writeQoS(null, that, packet)
         } else {
+          packet.writeCallback = () => {
+            packet.setDupIfNecessary(broker.persistence, that, cb)
+          }
           broker.persistence.outgoingUpdate(that, packet, writeQoS)
```

Comment:
This will trigger two `outgoingUpdate` one just after the other, I think we could do this in one